qiskit/providers/ibmq/ibmqfactory.py

```diff
@@ -444,36 +460,78 @@ def _check_api_version(credentials: Credentials) -> Dict[str, Union[bool, str]]:
 
-    def _initialize_providers(self, credentials: Credentials) -> None:
+    def _initialize_provider(self, credentials: Credentials,
+                             hub: Optional[str] = None,
+                             group: Optional[str] = None,
+                             project: Optional[str] = None) -> None:
         """Authenticate against IBM Quantum Experience and populate the providers.
 
         Args:
             credentials: Credentials for IBM Quantum Experience.
+            hub: Name of the hub to use.
+            group: Name of the group to use.
+            project: Name of the project to use.
+
+        Raises:
+            IBMQProviderError: if provider with ID `target_id` does not exist.
+
+        Returns:
+            AccountProvider: the provider
         """
+        self._credentials = credentials
+
         auth_client = AuthClient(credentials.token,
                                  credentials.base_url,
                                  **credentials.connection_parameters())
         service_urls = auth_client.current_service_urls()
-        user_hubs = auth_client.user_hubs()
+        user_hubs: List[Dict] = auth_client.user_hubs()
 
-        self._credentials = credentials
-        for hub_info in user_hubs:
-            # Build credentials.
-            provider_credentials = Credentials(
+        filters = []
+
+        if hub:
+            filters.append(lambda hgp: hgp['hub'] == hub)
+        if group:
+            filters.append(lambda hgp: hgp['group'] == group)
+        if project:
+            filters.append(lambda hgp: hgp['project'] == project)
+
+        # If all are none, load lazily
+        # note: this prevents a "load all"
+
+        if len(user_hubs) == 0:
+            raise IBMQProviderError('Provider cannot exist. The auth client provided has no hubs.')
+
+        if not (hub or group or project):
+            user_hubs = user_hubs[:1]  # Loads default only
+        else:
+            user_hubs = [uhub for uhub in user_hubs
+                         if all(f(uhub) for f in filters)]
+
+        # Successfully added a provider
+        success = False
+
+        for uhub in user_hubs:
+            # Build the credentials
+            creds = Credentials(
                 credentials.token,
                 access_token=auth_client.current_access_token(),
                 url=service_urls['http'],
                 websockets_url=service_urls['ws'],
                 proxies=credentials.proxies,
                 verify=credentials.verify,
                 services=service_urls.get('services', {}),
-                **hub_info, )
+                **uhub, )
 
             # Build the provider.
             try:
-                provider = AccountProvider(provider_credentials, self)
-                self._providers[provider_credentials.unique_id()] = provider
+                provider = AccountProvider(creds, self)
+                self._providers[creds.unique_id()] = provider
```

Comment:
So this would re-initialize the provider even if it has already been initialized?

Reply:
I think this is fixed. Added a check for the hub credentials already being initialized.